File: storage-common/src/main/java/io/github/solusmods/eternalcore/storage/impl/network/s2c/ClientAccess.java
package io.github.solusmods.eternalcore.storage.impl.network.s2c;


import io.github.solusmods.eternalcore.storage.api.StorageHolder;
import io.github.solusmods.eternalcore.storage.impl.CombinedStorage;
import net.minecraft.client.Minecraft;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.chunk.LevelChunk;
import org.jetbrains.annotations.Nullable;

public class ClientAccess {

    private ClientAccess() {
    }

    static void handle(SyncEntityStoragePayload packet) {
        Entity entity = getEntityFromId(packet.entityId());
        if (entity == null) return;
        handleUpdatePacket(entity, packet);
    }

    @Nullable
    static Entity getEntityFromId(int id) {
        // Early return if this is a self-update packet
        if (Minecraft.getInstance().player != null && Minecraft.getInstance().player.getId() == id)
            return Minecraft.getInstance().player;
        // Get entity from level
        Level level = Minecraft.getInstance().level;
        if (level == null) return null;
        return level.getEntity(id);
    }

    static void handle(SyncChunkStoragePayload packet) {
        Level level = Minecraft.getInstance().level;
        if (level == null) return;
        LevelChunk chunk = level.getChunk(packet.chunkPos().x, packet.chunkPos().z);
        handleUpdatePacket(chunk, packet);
    }

    static void handle(SyncWorldStoragePayload packet) {
        Level level = Minecraft.getInstance().level;
        if (level == null) return;
        handleUpdatePacket(level, packet);
    }


    static void handleUpdatePacket(StorageHolder holder, StorageSyncPayload packet) {
        if (packet.isUpdate()) {
            holder.eternalCraft$getCombinedStorage().handleUpdatePacket(packet.storageTag());
        } else {
            CombinedStorage updatedStorage = new CombinedStorage(holder);
            updatedStorage.load(packet.storageTag());
            holder.eternalCraft$setCombinedStorage(updatedStorage);
        }
    }
}

File: storage-common/src/main/kotlin/io/github/solusmods/eternalcore/storage/impl/network/s2c/ClientAccess.kt
package io.github.solusmods.eternalcore.storage.impl.network.s2c

import io.github.solusmods.eternalcore.storage.api.StorageHolder
import io.github.solusmods.eternalcore.storage.impl.CombinedStorage
import net.minecraft.client.Minecraft
import net.minecraft.world.entity.Entity

object ClientAccess {

    fun handle(packet: SyncEntityStoragePayload) {
        getEntityFromId(packet.entityId)?.let { entity ->
            handleUpdatePacket(entity, packet)
        }
    }

    fun getEntityFromId(id: Int): Entity? {
        val mc = Minecraft.getInstance()
        val player = mc.player
        if (player?.id == id) return player
        return mc.level?.getEntity(id)
    }

    fun handle(packet: SyncChunkStoragePayload) {
        Minecraft.getInstance().level?.let { level ->
            val chunk = level.getChunk(packet.chunkPos.x, packet.chunkPos.z)
            handleUpdatePacket(chunk, packet)
        }
    }

    fun handle(packet: SyncWorldStoragePayload) {
        Minecraft.getInstance().level?.let { level ->
            handleUpdatePacket(level, packet)
        }
    }

    fun handleUpdatePacket(holder: StorageHolder, packet: StorageSyncPayload) {
        val storage = if (packet.isUpdate) {
            holder.`eternalCore$getCombinedStorage`().apply {
                handleUpdatePacket(packet.storageTag)
            }
        } else {
            CombinedStorage(holder).apply {
                load(packet.storageTag)
                holder.`eternalCore$setCombinedStorage`(this)
            }
        }
    }
}
